src/utils/ScriptTransformer.ts

```diff
@@ -0,0 +1,154 @@
+    case "script":
+      return FileType.NS1;
+    default:
+      throw new Error(`Invalid extension: ${extension}. Filename: ${filename}.`);
+  }
+}
+
+export function getFileTypeFeature(fileType: FileType): FileTypeFeature {
+  const result: FileTypeFeature = {
+    isReact: false,
+    isTypeScript: false,
+  };
+  if (fileType === FileType.JSX || fileType === FileType.TSX) {
+    result.isReact = true;
+  }
+  if (fileType === FileType.TS || fileType === FileType.TSX) {
+    result.isTypeScript = true;
+  }
+  return result;
+}
+
+export function parseAST(code: string, fileType: FileType): AST {
+  const fileTypeFeature = getFileTypeFeature(fileType);
+  let ast: AST;
+  if (fileType === FileType.JS) {
+    ast = acorn.parse(code, { sourceType: "module", ecmaVersion: "latest" });
+  } else {
+    const plugins = [];
+    if (fileTypeFeature.isReact) {
+      plugins.push("jsx");
+    }
+    if (fileTypeFeature.isTypeScript) {
+      plugins.push("typescript");
+    }
+    ast = babel.packages.parser.parse(code, {
+      sourceType: "module",
+      ecmaVersion: "latest",
+      plugins: [["estree", { classFeatures: true }], ...plugins],
+    }).program;
+  }
+  return ast;
+}
+
+/**
+ * Simple module resolution algorithm:
+ * - Try each extension in validScriptExtensions
+ * - Return the first script found
+ */
+export function getModuleScript(moduleName: string, scripts: Map<ScriptFilePath, Script>): Script {
+  let script;
+  for (const extension of validScriptExtensions) {
+    const filename = resolveScriptFilePath(moduleName, root, extension);
+    if (!filename) {
+      throw new Error(`Invalid module: "${moduleName}"`);
+    }
+    script = scripts.get(filename);
+    if (script) {
+      break;
+    }
+  }
+  if (!script) {
+    throw new Error(`Invalid module: "${moduleName}"`);
+  }
+  return script;
+}
+
+export function transformScript(filename: string, code: string, fileType: FileType): string | null | undefined {
+  if (supportedFileTypes.every((v) => v !== fileType)) {
+    throw new Error(`Invalid file type: ${fileType}`);
+  }
+  const fileTypeFeature = getFileTypeFeature(fileType);
+  // This is only for testing. It will be removed after we decide between Babel and SWC.
+  if (globalThis.forceBabelTransform) {
+    const presets = [];
+    if (fileTypeFeature.isReact) {
+      presets.push("react");
+    }
+    if (fileTypeFeature.isTypeScript) {
+      presets.push("typescript");
+    }
+    return babel.transform(code, { filename: filename, presets: presets }).code;
+  }
+  let parserConfig: ParserConfig;
+  if (fileTypeFeature.isTypeScript) {
+    parserConfig = {
+      syntax: "typescript",
+    };
+    if (fileTypeFeature.isReact) {
+      parserConfig.tsx = true;
+    }
+  } else {
+    parserConfig = {
+      syntax: "ecmascript",
+    };
+    if (fileTypeFeature.isReact) {
+      parserConfig.jsx = true;
+    }
+  }
+  return transformSync(code, {
+    jsc: {
+      parser: parserConfig,
+      target: "es2020",
```

Comment:
Although it's usually not recommended, I feel like we might want to target esnext.

Reply:
I don't have a strong feeling about "es2020" vs "esnext", so I can change it if you insist. I choose it because that's when JS supports some really useful features (`?.`, `??`, `globalThis`, `dynamic import()`, etc.).

Reply:
The reason I suggested "esnext" is because it should prevent things from getting polyfilled and prevent complaints about bleeding-edge syntax, and in this context specifically (you are running a compiler *inside your browser*) it feels like you should let the browser give you errors about syntax it doesn't support (yet), rather than jsc.